File: 백준/Gold/16562. 친구비/친구비.py
# 특정 원소가 속한 집합을 찾기
def find_parent(parent, x):
    if parent[x] != x:
        parent[x] = find_parent(parent, parent[x])
    return parent[x]

# 두 원소가 속한 집합을 합치기
def union_parent(parent, a, b):
    a = find_parent(parent, a)
    b = find_parent(parent, b)
    if a<b:
        parent[b] = a
    else:
        parent[a] = b

n, m, k = map(int, input().split())
parent = [0] * (n+1)

for i in range(1, n+1):
    parent[i] = i

# 모든 간선을 담을 리스트와, 최종 비용을 담을 변수
edges = []
result = 0

money = list(map(int, input().split()))
# 모든 간선에 대한 정보를 입력받기
for _ in range(m):
    a, b = map(int, input().split())
    edges.append((0, a, b))

for i in range(n):
    edges.append((money[i], 0, i+1))

edges.sort()

for edge in edges:
    cost, a, b = edge
    if find_parent(parent, a)!=find_parent(parent, b):
        union_parent(parent, a, b)
        result += cost

if result<=k:
    print(result)
else:
    print("Oh no")

File: 백준/Gold/16562. 친구비/친구비.java
import java.util.*;
import java.io.*;

public class Main {

    static class Pair {
        int money;
        int friend1;
        int friend2;

        public Pair(int money, int friend1, int friend2){
            this.money = money;
            this.friend1 = friend1;
            this.friend2 = friend2;
        }

    }

    static int n, k;
    static int result = 0;
    static int[] parent;

    public static void main(String[] args) throws IOException{
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        int m = Integer.parseInt(st.nextToken());
        k = Integer.parseInt(st.nextToken());
        parent = new int[n+1];
        for (int i = 1; i<=n; i++) {
            parent[i] = i;
        }

        int[] money = new int[n];

        st = new StringTokenizer(br.readLine());
        for (int i = 0; i<n; i++) {
            money[i] = Integer.parseInt(st.nextToken());
        }

        ArrayList<Pair> edges = new ArrayList<>();

        for (int i = 0; i<m; i++) {
            st = new StringTokenizer(br.readLine());
            int friend1 = Integer.parseInt(st.nextToken());
            int friend2 = Integer.parseInt(st.nextToken());
            edges.add(new Pair(0, friend1, friend2));
        }

        for (int i = 0; i<n; i++) {
            edges.add(new Pair(money[i], 0, i+1));
        }

        edges.sort(Comparator.comparingInt(o -> o.money));

        for (Pair edge : edges) {
            int cost = edge.money;
            int a = edge.friend1;
            int b = edge.friend2;

            if (findParent(a) != findParent(b)) {
                unionParent(a, b);
                result += cost;
            }
        }

        System.out.println((result<=k) ? result : "Oh no");

    }

    public static int findParent(int x) {
        if (parent[x] != x) {
            parent[x] = findParent(parent[x]);
        }
        return parent[x];
    }

    public static void unionParent(int a, int b) {
        a = findParent(a);
        b = findParent(b);
        if (a < b) parent[b] = a;
        else parent[a] = b;
    }

}